File: WeeK_8.cpp
#include <iostream>
#include <string>

using namespace std;

class SinhVien {   // Định nghĩa lớp SinhVien
public:            // Khởi tạo các thuộc tính của lớp sinh viên
    string maSV;
    string hoTen;
    string gioiTinh;
    string ngaySinh;
    string diaChi;
    string lop;
    string khoa;
    SinhVien* next; // Con trỏ trỏ đến sinh viên tiếp theo
    SinhVien(string ma, string hoTen, string gioiTinh, string ngaySinh, string diaChi, string lop, string khoa)                   //Khởi tạo các thuộc tính của một đối tượng sinh viên mới
        : maSV(ma), hoTen(hoTen), gioiTinh(gioiTinh), ngaySinh(ngaySinh), diaChi(diaChi), lop(lop), khoa(khoa), next(nullptr) {}
};

class DanhSachSV {  // Định nghĩa lớp DanhSachSV
private:
    SinhVien* first; // Con trỏ trỏ đến sinh viên đầu tiên trong danh sách

public:
    DanhSachSV() : first(nullptr) {}   //Cho danh sách ban đầu rỗng

    bool isEmpty() {      // Kiểm tra xem danh sách có rỗng không, first trỏ đến nullptr -> true
        return first == nullptr;
    }

    void nhapSV() {  //Khởi tạo hàm nhập để điền thông tin của sinh viên
        string maSV, hoTen, gioiTinh, ngaySinh, diaChi, lop, khoa;
        cout << "Nhập mã số sinh viên: ";
        getline(cin, maSV);
        cout << "Nhập họ và tên sinh viên: ";
        getline(cin, hoTen);
        cout << "Nhập giới tính: ";
        getline(cin, gioiTinh);
        cout << "Nhập ngày tháng năm sinh: ";
        getline(cin, ngaySinh);
        cout << "Nhập nơi ở hiện tại: ";
        getline(cin, diaChi);
        cout << "Nhập lớp sinh viên: ";
        getline(cin, lop);
        cout << "Nhập ngành học của sinh viên: ";
        getline(cin, khoa);

        SinhVien* newSV = new SinhVien(maSV, hoTen, gioiTinh, ngaySinh, diaChi, lop, khoa);         // Tạo một đối tượng sinh viên mới với thông tin vừa nhập
        newSV->next = first;
        first = newSV;  //thêm sinh viên mới vào đầu danh sách
    }

    void Bubble_Sort() {      //Sắp xếp mssv tăng dần bằng thuật toán nổi bọt
        if (isEmpty()) return;
        SinhVien* current = first;
        while (current) {
            SinhVien* next_node = current->next;
            while (next_node) {
                if (current->maSV > next_node->maSV) {   //nếu mssv của sv hiện tại lớn hơn sv tiếp theo thì hoán đổi các thuộc tính
                    swap(current->maSV, next_node->maSV);
                    swap(current->hoTen, next_node->hoTen);
                    swap(current->gioiTinh, next_node->gioiTinh);
                    swap(current->ngaySinh, next_node->ngaySinh);
                    swap(current->diaChi, next_node->diaChi);
                    swap(current->lop, next_node->lop);
                    swap(current->khoa, next_node->khoa);
                }
                next_node = next_node->next;
            }
            current = current->next;
        }
    }

    void timSV(const string& ngaySinh) {   //tìm sv theo ngày sinh
        SinhVien* current = first;
        bool found = false;
        while (current) {
            if (current->ngaySinh == ngaySinh) {
                cout << "Mã SV: " << current->maSV << ", Họ tên: " << current->hoTen << endl;
                found = true;
            }
            current = current->next;
        }
        if (!found) {
            cout << "Không tìm thấy sinh viên cùng ngày sinh" << endl;
        }
    }

    void xoaSV(const string& ngaySinh) {   //xóa sv theo ngày sinh
        SinhVien* current = first;
        SinhVien* prev = nullptr;
        while (current) {
            if (current->ngaySinh == ngaySinh) {
                if (prev) {
                    prev->next = current->next;
                } else {
                    first = current->next;
                }
                delete current; // Giải phóng bộ nhớ
                current = (prev) ? prev->next : first;
            } else {
                prev = current;
                current = current->next;
            }
        }
    }

    void inDanhSach() {    //in danh sách sinh viên
        SinhVien* current = first;
        while (current) {
            cout << "Mã SV: " << current->maSV << ", Họ tên: " << current->hoTen << ", Giới tính: " << current->gioiTinh
                 << ", Ngày sinh: " << current->ngaySinh << ", Địa chỉ: " << current->diaChi
                 << ", Lớp: " << current->lop << ", Khoa: " << current->khoa << endl;
            current = current->next;
        }
    }

    // Hàm hủy để giải phóng bộ nhớ
    ~DanhSachSV() {
        while (first) {
            SinhVien* temp = first;
            first = first->next;
            delete temp;
        }
    }
};

int main() {
    DanhSachSV ds;
    for (int i = 0; i < 5; ++i) {      // Nhập một số sinh viên
        ds.nhapSV();
    }
    ds.Bubble_Sort();
    cout << "Danh sách sinh viên sau khi sắp xếp:" << endl;
    ds.inDanhSach();
    string ngaySinh;
    cout << "Nhập ngày sinh cần tìm: ";
    getline(cin, ngaySinh);
    ds.timSV(ngaySinh);  //tìm sv theo ngày sinh

    ds.xoaSV(ngaySinh);  //xóa sv theo ngày sinh

    cout << "Danh sách sinh viên sau khi xóa:" << endl;   //in danh sách sau khi xóa
    ds.inDanhSach();
    return 0;
}

File: WeeK_8.py
class SinhVien:                                                                #Khai báo lớp sinh viên với các thuộc tính đề bài đã cho
    def __init__(self, maSV, hoTen, gioiTinh, ngaySinh, diaChi, lop, khoa):    #Khởi tạo các thuộc tính của đối tượng sinh viên
        self.maSV = maSV                                                       
        self.hoTen = hoTen                                                     
        self.gioiTinh = gioiTinh                                               
        self.ngaySinh = ngaySinh                                                    
        self.diaChi = diaChi                                                   
        self.lop = lop                                                          
        self.khoa = khoa                                                       
        self.next = None                                                       #Con trỏ trỏ đến sinh viên tiếp theo trong danh sách

class DanhSachSV:                                                              
    def __init__(self):                                                        
        self.first = None                                                      #Khởi tạo thuộc tính first để lưu địa chỉ của sinh viên đầu tiên trong danh sách, nếu danh sách rỗng thì địa chỉ là None

    def isEmpty(self):                                                         #Kiểm tra xem danh sách có rỗng không
        return self.first == None                                              #Trả về True -> danh sách rỗng và false ngược lại

    def nhapSV(self):
        maSV = input("Nhập mã số sinh viên: ")
        hoTen = input("Nhập họ và tên sinh viên: ")
        gioiTinh = input("Nhập giới tính: ")
        ngaySinh = input("Nhập ngày tháng năm sinh: ")
        diaChi = input("Nhập nơi ở hiện tại: ")
        lop = input("Nhập lớp sinh viên: ")
        khoa = input("Nhập ngành học của sinh viên: ")
        newSV = SinhVien(maSV, hoTen, gioiTinh, ngaySinh, diaChi, lop, khoa)   #Tạo một đối tượng sinh viên mới với thông tin người dùng nhập
        newSV.next = self.first                                                #Đặt con trỏ của sinh viên mới trỏ đến sinh viên hiện tại
        self.first = newSV

    def Bubble_sort(self):                                                     #Dùng thuật toán sắp xếp nổi bọt để sắp xếp mssv từ bé đến lớn
        current = self.first
        while current:
            next_node = current.next
            while next_node:
                if current.maSV > next_node.maSV:
                    current.maSV, next_node.maSV = next_node.maSV, current.maSV
                next_node = next_node.next
            current = current.next

    def themSV(self, svMoi):
        # Tìm vị trí thích hợp để chèn
        current = self.head
        prev = None
        while current and current.maSV < svMoi.maSV:
            prev = current
            current = current.next
        
        # Chèn sinh viên mới
        svMoi.next = current
        if prev:
            prev.next = svMoi
        else:
            self.head = svMoi

    def timSVTheoNgaySinh(self, ngaySinh):
        current = self.head
        while current:
            if current.ngaySinh == ngaySinh:
                print(f"Mã SV: {current.maSV}, Họ tên: {current.hoTen}")
            current = current.next
        if current == None:
            print("Không tìm thấy sinh viên cùng ngày sinh")

    def xoaSVTheoNgaySinh(self, ngaySinh):
        current = self.head
        prev = None
        while current:
            if current.ngaySinh == ngaySinh:
                if prev:
                    prev.next = current.next
                else:
                    self.head = current.next
            else:
                prev = current
            current = current.next

    def inDanhSach(self):
        current = self.head
        while current:
            print(f"Mã SV: {current.maSV}, Họ tên: {current.hoTen}, ...")
            current = current.next

if __name__ == "__main__":
    ds = DanhSachSV()

    # Nhập một số sinh viên
    for _ in range(3):
        ds.nhapSV()

    # Sắp xếp danh sách
    ds.sapXepTangDan()

    # In danh sách sau khi sắp xếp
    ds.inDanhSach()

    # Tìm sinh viên theo ngày sinh
    ngaySinh = input("Nhập ngày sinh cần tìm: ")
    ds.timSVTheoNgaySinh(ngaySinh)

    # Xóa sinh viên theo ngày sinh
    ds.xoaSVTheoNgaySinh(ngaySinh)

    # In danh sách sau khi xóa
    ds.inDanhSach()